### DISCUSSION TITLE
node PKCE

### DISCUSSION THREAD

**Discussion by anthonygiuliano** [jsforce/jsforce#1503] [Q&A] [answered]
I can't seem to figure out how to use PKCE in the node module. It looks like it's implemented on the command line but I don't see it in the Connection.js or Oauth.js you get from importing jsforce in Javascript. What am I missing?

**Comment by AllanOricil** [comment, marked as answer]
This will be solved once oauth module is refactored to use openid-client

https://github.com/jsforce/jsforce/discussions/1744


To use PKCE flow, first make your Connected App public by not requiring secret (I think there is a tick box you must not check). Then create oauth2.Connection class passing your ClientId and useVerifier = true. JSforce lib will generate both code_verifier and code_challenge for you, and send to Salesforce auth server. Then use `getAuthorizationUrl()` method and send the user to that URL. What exactly you were not able to do?


```js
import { OAuth2 } from 'jsforce';

const oauth2 = new OAuth2({
clientId: "...",
redirectUri: "...",
useVerifier: true,
});


// NOTE: get an authorization url only when user clicks on the login button of you app
const authUrl = oauth2.getAuthorizationUrl({ scope: "openid email profile" });

// NOTE: then use this authUrl to redirect the User
```

The `redirecUri` must be a route that your app will use to receive the "code" which you have to use to get tokens. It must be configured in the connected app. You can exchange the code by tokens in the client or in your server, if you have one. In the server you can use cookies to store your tokens (httpOnly, secure, strict/lax). If you have a redis instance, you can generate a sid and store your tokens in redis, then send this sid as a cookie to your client. This last step isn't really required if you apply csrf, cors, content-security policies and cookies best practices, use HTTPS all the time in your client and server. Also, configure tokens to expire soon, during logout invalidate tokens, and force the user to sign in again every now and then invalidating the refresh token.
